{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get all info for each cell in scRNA-seq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_info_df = pd.read_csv('../data/HN_patient_specific/AS_et_al_cellInfo2.csv', index_col=0)\n",
    "\n",
    "sra_run_df = pd.read_csv('../data/HN_patient_specific/AS_et_at_SraRunTable.txt', index_col='Run')\n",
    "\n",
    "sra_summary_df = pd.read_csv('../data/HN_patient_specific/AS_et_al_sra_result.csv', index_col='Sample Accession')\n",
    "# Drop Instrument as it is already in sra_run_df\n",
    "sra_summary_df = sra_summary_df.drop('Instrument', axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1241, 5) 1241\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cellLines</th>\n",
       "      <th>CA_site</th>\n",
       "      <th>Identity</th>\n",
       "      <th>identity2</th>\n",
       "      <th>Cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RHC6558</th>\n",
       "      <td>HN148M</td>\n",
       "      <td>Metastasis</td>\n",
       "      <td>3</td>\n",
       "      <td>32</td>\n",
       "      <td>D3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RHC6559</th>\n",
       "      <td>HN148M</td>\n",
       "      <td>Metastasis</td>\n",
       "      <td>3</td>\n",
       "      <td>32</td>\n",
       "      <td>D3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RHC6560</th>\n",
       "      <td>HN148M</td>\n",
       "      <td>Metastasis</td>\n",
       "      <td>3</td>\n",
       "      <td>32</td>\n",
       "      <td>D3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RHC6561</th>\n",
       "      <td>HN148M</td>\n",
       "      <td>Metastasis</td>\n",
       "      <td>3</td>\n",
       "      <td>32</td>\n",
       "      <td>D3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RHC6562</th>\n",
       "      <td>HN148M</td>\n",
       "      <td>Metastasis</td>\n",
       "      <td>3</td>\n",
       "      <td>32</td>\n",
       "      <td>D3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        cellLines     CA_site  Identity  identity2 Cluster\n",
       "RHC6558    HN148M  Metastasis         3         32      D3\n",
       "RHC6559    HN148M  Metastasis         3         32      D3\n",
       "RHC6560    HN148M  Metastasis         3         32      D3\n",
       "RHC6561    HN148M  Metastasis         3         32      D3\n",
       "RHC6562    HN148M  Metastasis         3         32      D3"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print (cluster_info_df.shape, len(set(cluster_info_df.index)))\n",
    "cluster_info_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Run file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1527, 28) 1527\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Run</th>\n",
       "      <th>SRR7622080</th>\n",
       "      <th>SRR7622081</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Assay Type</th>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>RNA-Seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AvgSpotLen</th>\n",
       "      <td>100</td>\n",
       "      <td>150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BioProject</th>\n",
       "      <td>PRJNA483471</td>\n",
       "      <td>PRJNA483471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BioSample</th>\n",
       "      <td>SAMN09740929</td>\n",
       "      <td>SAMN09740929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell_type</th>\n",
       "      <td>Primary</td>\n",
       "      <td>Primary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Center Name</th>\n",
       "      <td>GEO</td>\n",
       "      <td>GEO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Consent</th>\n",
       "      <td>public</td>\n",
       "      <td>public</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DATASTORE filetype</th>\n",
       "      <td>sra</td>\n",
       "      <td>sra</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DATASTORE provider</th>\n",
       "      <td>gs,ncbi,s3</td>\n",
       "      <td>gs,ncbi,s3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DATASTORE region</th>\n",
       "      <td>gs.US,ncbi.public,s3.us-east-1</td>\n",
       "      <td>gs.US,ncbi.public,s3.us-east-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>disease_state</th>\n",
       "      <td>oral squamous cell carcinoma (OSCC)</td>\n",
       "      <td>oral squamous cell carcinoma (OSCC)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Experiment</th>\n",
       "      <td>SRX4486170</td>\n",
       "      <td>SRX4486170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GEO_Accession</th>\n",
       "      <td>GSM3313660</td>\n",
       "      <td>GSM3313660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Instrument</th>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LibraryLayout</th>\n",
       "      <td>SINGLE</td>\n",
       "      <td>SINGLE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LibrarySelection</th>\n",
       "      <td>cDNA</td>\n",
       "      <td>cDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LibrarySource</th>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MBases</th>\n",
       "      <td>237</td>\n",
       "      <td>727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MBytes</th>\n",
       "      <td>107</td>\n",
       "      <td>293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Organism</th>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>passages</th>\n",
       "      <td>13-15</td>\n",
       "      <td>13-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Platform</th>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>ILLUMINA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ReleaseDate</th>\n",
       "      <td>2018-11-23T00:00:00Z</td>\n",
       "      <td>2018-11-23T00:00:00Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>response_to_cisplatin</th>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_acc</th>\n",
       "      <td>SRS3610332</td>\n",
       "      <td>SRS3610332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample Name</th>\n",
       "      <td>GSM3313660</td>\n",
       "      <td>GSM3313660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>source_name</th>\n",
       "      <td>HN137P</td>\n",
       "      <td>HN137P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRA Study</th>\n",
       "      <td>SRP155751</td>\n",
       "      <td>SRP155751</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Run                                             SRR7622080  \\\n",
       "Assay Type                                         RNA-Seq   \n",
       "AvgSpotLen                                             100   \n",
       "BioProject                                     PRJNA483471   \n",
       "BioSample                                     SAMN09740929   \n",
       "Cell_type                                          Primary   \n",
       "Center Name                                            GEO   \n",
       "Consent                                             public   \n",
       "DATASTORE filetype                                     sra   \n",
       "DATASTORE provider                              gs,ncbi,s3   \n",
       "DATASTORE region            gs.US,ncbi.public,s3.us-east-1   \n",
       "disease_state          oral squamous cell carcinoma (OSCC)   \n",
       "Experiment                                      SRX4486170   \n",
       "GEO_Accession                                   GSM3313660   \n",
       "Instrument                             Illumina HiSeq 2500   \n",
       "LibraryLayout                                       SINGLE   \n",
       "LibrarySelection                                      cDNA   \n",
       "LibrarySource                               TRANSCRIPTOMIC   \n",
       "MBases                                                 237   \n",
       "MBytes                                                 107   \n",
       "Organism                                      Homo sapiens   \n",
       "passages                                             13-15   \n",
       "Platform                                          ILLUMINA   \n",
       "ReleaseDate                           2018-11-23T00:00:00Z   \n",
       "response_to_cisplatin                            Sensitive   \n",
       "sample_acc                                      SRS3610332   \n",
       "Sample Name                                     GSM3313660   \n",
       "source_name                                         HN137P   \n",
       "SRA Study                                        SRP155751   \n",
       "\n",
       "Run                                             SRR7622081  \n",
       "Assay Type                                         RNA-Seq  \n",
       "AvgSpotLen                                             150  \n",
       "BioProject                                     PRJNA483471  \n",
       "BioSample                                     SAMN09740929  \n",
       "Cell_type                                          Primary  \n",
       "Center Name                                            GEO  \n",
       "Consent                                             public  \n",
       "DATASTORE filetype                                     sra  \n",
       "DATASTORE provider                              gs,ncbi,s3  \n",
       "DATASTORE region            gs.US,ncbi.public,s3.us-east-1  \n",
       "disease_state          oral squamous cell carcinoma (OSCC)  \n",
       "Experiment                                      SRX4486170  \n",
       "GEO_Accession                                   GSM3313660  \n",
       "Instrument                             Illumina HiSeq 2500  \n",
       "LibraryLayout                                       SINGLE  \n",
       "LibrarySelection                                      cDNA  \n",
       "LibrarySource                               TRANSCRIPTOMIC  \n",
       "MBases                                                 727  \n",
       "MBytes                                                 293  \n",
       "Organism                                      Homo sapiens  \n",
       "passages                                             13-15  \n",
       "Platform                                          ILLUMINA  \n",
       "ReleaseDate                           2018-11-23T00:00:00Z  \n",
       "response_to_cisplatin                            Sensitive  \n",
       "sample_acc                                      SRS3610332  \n",
       "Sample Name                                     GSM3313660  \n",
       "source_name                                         HN137P  \n",
       "SRA Study                                        SRP155751  "
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print (sra_run_df.shape, len(set(sra_run_df.index)))\n",
    "# sra_run_df.head(1).T\n",
    "sra_run_df[sra_run_df['GEO_Accession'].isin(['GSM3313660'])].T"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Drop smaller library size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "sra_run_df = sra_run_df.sort_values(by=['GEO_Accession', 'MBases'], ascending=[True, False])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GEO_Accession</th>\n",
       "      <th>MBases</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Run</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR7622081</th>\n",
       "      <td>GSM3313660</td>\n",
       "      <td>727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR7622080</th>\n",
       "      <td>GSM3313660</td>\n",
       "      <td>237</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           GEO_Accession  MBases\n",
       "Run                             \n",
       "SRR7622081    GSM3313660     727\n",
       "SRR7622080    GSM3313660     237"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sra_run_df[sra_run_df['GEO_Accession']=='GSM3313660'][['GEO_Accession', 'MBases']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "sra_run_df = sra_run_df.drop_duplicates(subset=['GEO_Accession'], keep='first')\n",
    "sra_run_df = sra_run_df.reset_index().set_index('sample_acc')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Summary file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "sra_summary_df.loc[:, 'Cell ID'] = [c.split(' ')[1] for c in sra_summary_df['Experiment Title']]\n",
    "sra_summary_df.loc[:, 'GEO_Accession'] = [c.split(': ')[0] for c in sra_summary_df['Experiment Title']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1431, 17) 1431\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Sample Accession</th>\n",
       "      <th>SRS3611000</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Experiment Accession</th>\n",
       "      <td>SRX4486839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Experiment Title</th>\n",
       "      <td>GSM3314329: RHC6653 (scRNA-seq); Homo sapiens;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Organism Name</th>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Submitter</th>\n",
       "      <td>GEO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Study Accession</th>\n",
       "      <td>SRP155751</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Study Title</th>\n",
       "      <td>Single cell RNA-seq profiles of primary, metas...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample Title</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total Size, Mb</th>\n",
       "      <td>1.42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total RUNs</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total Spots</th>\n",
       "      <td>32549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total Bases</th>\n",
       "      <td>3254900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Library Name</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Library Strategy</th>\n",
       "      <td>RNA-Seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Library Source</th>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Library Selection</th>\n",
       "      <td>cDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell ID</th>\n",
       "      <td>RHC6653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GEO_Accession</th>\n",
       "      <td>GSM3314329</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Sample Accession                                             SRS3611000\n",
       "Experiment Accession                                         SRX4486839\n",
       "Experiment Title      GSM3314329: RHC6653 (scRNA-seq); Homo sapiens;...\n",
       "Organism Name                                              Homo sapiens\n",
       "Submitter                                                           GEO\n",
       "Study Accession                                               SRP155751\n",
       "Study Title           Single cell RNA-seq profiles of primary, metas...\n",
       "Sample Title                                                        NaN\n",
       "Total Size, Mb                                                     1.42\n",
       "Total RUNs                                                            1\n",
       "Total Spots                                                       32549\n",
       "Total Bases                                                     3254900\n",
       "Library Name                                                        NaN\n",
       "Library Strategy                                                RNA-Seq\n",
       "Library Source                                           TRANSCRIPTOMIC\n",
       "Library Selection                                                  cDNA\n",
       "Cell ID                                                         RHC6653\n",
       "GEO_Accession                                                GSM3314329"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print (sra_summary_df.shape, len(set(sra_summary_df.index)))\n",
    "sra_summary_df.head(1).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [],
   "source": [
    "sra_df = pd.merge(sra_run_df, sra_summary_df, left_index=True, right_index=True)\n",
    "sra_df = sra_df.reset_index()\n",
    "sra_df.columns = ['Sample Accession'] + list(sra_df.columns[1:])\n",
    "sra_df = sra_df.set_index('Cell ID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1431 1431\n"
     ]
    }
   ],
   "source": [
    "print (len(set(sra_df.index)), len(sra_df.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_meta_df = pd.merge(sra_df, cluster_info_df, how='left', left_index=True, right_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_meta_df.to_csv('../data/HN_patient_specific/AS_et_al_cell_metadata.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": false,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
